Windows x64 build

Hi,

I'm trying to build editorconfig/editorconfig-notepad-plus-plus#9 for x64.
Which build options should i set to build pcre?

Regards
Michael

If i compile with md, i got linker errors for `editorconfig.dll`:

```
ec_glob.obj : error LNK2019: unresolved external symbol __imp_pcre_compile referenced in function ec_glob
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,3 +1,8 @@
+v0.12.4
+
+- Add Windows build scripts.
+- Updated AppVeyor to use new scripts.
+
 v0.12.3
 
 - Bump required minimum cmake version to 2.8.12.

--- a/INSTALL.md
+++ b/INSTALL.md
@@ -89,9 +89,71 @@ PREFIX/lib as one of the library searching directory on UNIX/Linux to make sure
 that source files could be linked to the libraries and executables depending on
 these libraries could be executed properly.
 
-On Windows, via Developer Command Prompt for Visual Studio:
+## Build on Windows
+Requirements on Windows are [Visual Studio] 2013, 2015 or 2017, [cmake] 2.8.12 or higher and Powershell 3 or higher. For Visual Studio the community edition is sufficient, but the [C++ workload](https://docs.microsoft.com/en-us/cpp/build/vscpp-step-0-installation?view=vs-2017) is required.
 
-    msbuild all_build.vcxproj /p:Configuration=Release
+    Non-static build is currently not supported.
 
-[cmake]: http://www.cmake.org
-[PCRE2]: http://pcre.org
+### Download pcre2
+You have to download and extract the pcre2 sources after checkout.
+
+```powershell
+~> ./init.ps1 [-pcre 10.32]
+```
+
+Arguments:
+
+    -pcre  Optional; pcre2 version to download.
+
+### Build library
+To build pcre2 and editorconfig core library in one step use the `-init` and `-install` arguments at the same time.
+
+```powershell
+~> ./build.ps1 -init -install
+```
+
+The `-init` argument will generate the required cmake build files for Visual Studio. This is required after initial checkout or `CMakeLists.txt` changes.
+The `-install` argument will put the binaries to a location (`bin/$(ARCH)-static/build`) that the editorconfig project can find and link the library. This folder can be used to distribute the build binaries.
+For the other arguments please see below.
+
+```powershell
+~> ./build.ps1 [-proj all | core | pcre2] [-init] [-install] [-vsver 15 | 14 | 12] [-arch x64 | x86] [-config Release | Debug]
+```
+
+Arguments:
+
+    -proj Optional; Project to build.
+    -init Optional; (Re)Generate cmake build files, required first time or on `CMakeLists.txt` changes.
+    -install Optional; Install to `bin/$(ARCH)-static/build` folder.
+    -vsver Optional; Visual Studio version (major version number) to use.
+    -arch Optional; Architecture to build for.
+    -config Optional; Debug or release build.
+
+#### Build pcre2 library
+```powershell
+~> ./build.ps1 -proj pcre2 -init -install
+```
+
+#### Build editorconfig core library
+
+```powershell
+~> ./build.ps1 -proj core -init
+```
+
+
+### Run tests
+On Windows the test `utf_8_char` is disabled.
+
+```powershell
+~> ./test.ps1 [-arch x64 | x86] [-config Release | Debug]
+```
+
+Arguments:
+
+    -arch Optional; Architecture to build for.
+    -config Optional; Debug or release build.
+ 
+
+[cmake]: https://cmake.org
+[PCRE2]: https://pcre.org/
+[Visual Studio]: https://visualstudio.microsoft.com

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -2,41 +2,28 @@ image:
   - Visual Studio 2017
   
 environment:
-  CMAKE_GENERATOR: 'Visual Studio 15 2017 Win64'
   CONFIG: Release
-  ARCH: x64
-  CMAKE_INSTALL_PREFIX: ..\build
-  PCRE2_DEST: C:\projects\bin\%ARCH%-static\pcre2
-  CORE_DEST: C:\projects\bin\%ARCH%-static\core
+  VSVER: 15
+  matrix:
+    - ARCH: x86
+    - ARCH: x64
 
 install:
   - cmake --version
-  - git submodule init
-  - git submodule update
-  - cd C:\projects
+  - git submodule update --init
   # Download PCRE2 sources
-  - curl -o pcre2.zip https://ftp.pcre.org/pub/pcre/pcre2-10.32.zip
-  - 7z x -y pcre2.zip > nul
-  - rename pcre2-10.32 pcre2
+  - ps: ./init.ps1 -pcre 10.32
   # Build and install PCRE2
-  - mkdir %PCRE2_DEST%
-  - cd %PCRE2_DEST%
-  - cmake -G "%CMAKE_GENERATOR%" -DCMAKE_INSTALL_PREFIX="%CMAKE_INSTALL_PREFIX%" -DPCRE2_STATIC_RUNTIME="ON" -DBUILD_SHARED_LIBS="OFF" -DPCRE2_BUILD_PCRE2GREP="OFF" -DPCRE2_BUILD_TESTS="OFF" C:\projects\pcre2
-  - cd %PCRE2_DEST%
-  - cmake --build %PCRE2_DEST% --target install -- /p:Configuration=%CONFIG%
+  - ps: ./build.ps1 -proj pcre2 -config $env:CONFIG -arch $env:ARCH -vsver $env:VSVER -init -install
 
 build_script:
   # Build and install editorconfig-core-c
-  - mkdir %CORE_DEST%
-  - cd %CORE_DEST%
-  - cmake -G "%CMAKE_GENERATOR%" -DCMAKE_INSTALL_PREFIX="%CMAKE_INSTALL_PREFIX%" -DMSVC_MD="OFF" -DPCRE2_STATIC="ON" C:\projects\editorconfig-core-c
-  - cmake --build %CORE_DEST% --target install -- /p:Configuration=%CONFIG%
+  - ps: ./build.ps1 -proj core -config $env:CONFIG -arch $env:ARCH -vsver $env:VSVER -init -install
 
 test_script:
   # Run the core tests
-  - cd %CORE_DEST%
-  - ctest -E utf_8_char -VV --output-on-failure .
+  - ps: ./test.ps1 -proj core -config $env:CONFIG -arch $env:ARCH
 
 on_failure:
   - echo. && echo. && echo. && echo "Tests for editorconfig-core-c failed. Here is their detailed output:" && echo. && echo. && echo.
-  - type C:\projects\bin\x64-static\core\Testing\Temporary\LastTest.log
+  - type .\bin\x64-static\core\Testing\Temporary\LastTest.log

--- a/build.ps1
+++ b/build.ps1
@@ -1,0 +1,137 @@
+param(
+    [ValidateSet("pcre2","core", "all")]
+    [string] $proj = "all",
+
+    [switch] $init,
+
+    [switch] $install,
+    
+    [ValidateSet("ON","OFF")]
+    [string] $static = "ON",
+
+    [ValidateSet("Debug","Release")]
+    [string] $config = "Release",
+
+    [ValidateSet("x86","x64")]
+    [string] $arch = "x64",
+    
+
+    [ValidateSet("15","14","12")]
+    [int] $vsver = 15
+)
+
+$ErrorActionPreference="stop"
+
+function exec
+{
+    param
+    (
+        [ScriptBlock] $ScriptBlock,
+        [int[]] $AllowedExitCodes = @(0)
+    )
+ 
+    $backupErrorActionPreference = $script:ErrorActionPreference
+ 
+    $script:ErrorActionPreference = "Continue"
+    try
+    {
+        & $ScriptBlock
+        if ($AllowedExitCodes -notcontains $LASTEXITCODE)
+        {
+            throw "Execution failed with exit code $LASTEXITCODE"
+        }
+    }
+    finally
+    {
+        $script:ErrorActionPreference = $backupErrorActionPreference
+    }
+}
+
+if ($proj -eq "all"){
+    .\build.ps1 -proj pcre2 -init:$init -install:$install -arch $arch -config $config -static $static
+    .\build.ps1 -proj core -init:$init -install:$install -arch $arch -config $config -static $static
+    return
+}
+
+$PREFIX="../build"
+
+$dest = "bin"
+
+if ((Test-Path $dest) -ne $true){
+    throw "Missing build path! Used init?"
+}
+
+$dest += "\$arch"
+
+if($static -eq "ON"){
+    $dest += "-static"
+}
+
+$dest += "\$proj"
+
+if ($init) {
+    "Generating $proj build files" | Write-Host -ForegroundColor DarkGreen
+
+    $gen = "Visual Studio "
+
+    switch ($vsver) {
+        15 {
+            $gen += "15 2017"
+        }
+        14 {
+            $gen += "14 2015"
+        }
+        12 {
+            $gen += "12 2013"
+        }
+        default {
+            throw "Visual Studio version $vsver not supported!"
+        }
+    }
+
+    if($arch -eq "x64"){
+        $gen += " Win64"
+    }
+
+    mkdir $dest -ErrorAction SilentlyContinue | Out-Null
+    Push-Location $dest
+    try {
+        switch ($proj) {
+            pcre2 {
+                $BUILD_SHARED_LIBS = "ON"
+                if ($static -eq "ON"){ $BUILD_SHARED_LIBS = "OFF"}
+                exec { cmake -G "$gen" -DCMAKE_INSTALL_PREFIX="$PREFIX" -DPCRE2_STATIC_RUNTIME="$static" `
+                    -DBUILD_SHARED_LIBS="$BUILD_SHARED_LIBS" -DPCRE2_BUILD_PCRE2GREP="OFF" -DPCRE2_BUILD_TESTS="OFF" `
+                    "../../pcre2" }
+            }
+            core {
+                $MSVC_MD = "ON"
+                if ($static -eq "ON"){ $MSVC_MD = "OFF"}
+                exec { cmake -G "$gen" -DCMAKE_INSTALL_PREFIX="$PREFIX" -DMSVC_MD="$MSVC_MD" -DPCRE2_STATIC="$static" `
+                    "../../../." }
+            }
+        }
+    } finally {
+        Pop-Location
+    }
+}
+
+if ((Test-Path $dest) -ne $true){
+    throw "Missing build path! Used init?"
+}
+
+"Compiling $proj" | Write-Host -ForegroundColor DarkGreen
+
+exec { cmake --build $dest `-- /p:Configuration=$config }
+
+if ($install) {
+    "Installing $proj" | Write-Host -ForegroundColor DarkGreen
+    switch ($proj) {
+        pcre2 {
+            exec { cmake --build $dest --target install `-- /p:Configuration=$config }
+        }
+        core {
+            exec { cmake --build $dest --target install `-- /p:Configuration=$config }
+        }
+    }
+}

--- a/init.ps1
+++ b/init.ps1
@@ -1,0 +1,30 @@
+param(
+    $pcre="10.32"
+)
+
+$ErrorActionPreference="Stop"
+$dest = "bin"
+
+if (Test-Path $dest){
+    Remove-Item $dest -Recurse -Force -Confirm:$false
+}
+
+New-Item $dest -ItemType Directory -ErrorAction SilentlyContinue | Out-Null
+
+#####################################################################
+# pcre
+#####################################################################
+$url = "https://ftp.pcre.org/pub/pcre/pcre2-$($pcre).zip"
+$output = "$dest\pcre-$($pcre).zip"
+
+"Downloading pcre2 v$pcre sources" | Write-Host -ForegroundColor DarkGreen
+Start-BitsTransfer -Source $url -Destination $output
+
+"Extracting pcre2 v$pcre sources" | Write-Host -ForegroundColor DarkGreen
+Expand-Archive -Path $output -DestinationPath $dest
+Rename-Item -Path "$dest\pcre2-$($pcre)" -NewName "pcre2"
+
+if(Test-Path "pcre-$($pcre).patch") {
+    "Applying pcre patches" | Write-Host -ForegroundColor DarkGreen
+    .\patch.ps1 -patch "pcre-$($pcre).patch" -path $dest\pcre2
+}

--- a/patch.ps1
+++ b/patch.ps1
@@ -1,0 +1,42 @@
+param(
+    $path,
+    $patch
+)
+
+$ErrorActionPreference="Stop"
+$dest = "tools"
+
+if ((Test-Path $dest) -ne $true){
+    New-Item $dest -ItemType Directory | Out-Null
+}
+
+#####################################################################
+# patch
+#####################################################################
+if (-ne Test-Path "$dest\bin\patch.exe") {
+    $url = "http://gnuwin32.sourceforge.net/downlinks/patch-bin-zip.php"
+    $output = "$dest\patch.zip"
+
+    "Downloading patch binary" | Write-Host -ForegroundColor DarkGreen
+    Start-BitsTransfer -Source $url -Destination $output
+
+    "Extracting patch binary" | Write-Host -ForegroundColor DarkGreen
+    Expand-Archive -Path $output -DestinationPath $dest
+@"
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+<trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
+    <security>
+    <requestedPrivileges>
+        <!-- Make sure that UAC believes
+        that it does not require administrative privilege -->
+        <requestedExecutionLevel level="asInvoker" uiAccess="false"/>
+    </requestedPrivileges>
+    </security>
+</trustInfo>
+</assembly>
+"@ | Out-File -FilePath $dest\bin\patch.exe.manifest -Encoding utf8
+    (Get-ChildItem $dest\bin\patch.exe).LastWriteTime = Get-Date
+}
+
+Get-Content $patch | & $dest\bin\patch.exe -d $path

--- a/test.ps1
+++ b/test.ps1
@@ -1,0 +1,46 @@
+param(
+    [ValidateSet("pcre2","core")]
+    [string] $proj = "core",
+    
+    [ValidateSet("ON","OFF")]
+    [string] $static = "ON",
+
+    [ValidateSet("Debug","Release")]
+    [string] $config = "Release",
+
+    [ValidateSet("x86","x64")]
+    [string] $arch = "x64"
+)
+
+$ErrorActionPreference="stop"
+
+
+$dest = "bin"
+
+if ((Test-Path $dest) -ne $true){
+    throw "Missing build path! Used init?"
+}
+
+$dest += "\$arch"
+
+if($static -eq "ON"){
+    $dest += "-static"
+}
+
+
+$bin = "$dest\build\bin"
+$dest += "\$proj"
+
+if ((Test-Path $dest) -ne $true){
+    throw "Missing build path! Used init?"
+}
+
+$env:Path += ";" + (Resolve-Path $bin)
+
+"Testing $proj" | Write-Host -ForegroundColor DarkGreen
+Push-Location $dest
+try {
+    ctest -E utf_8_char -VV --output-on-failure . -C $config
+} finally {
+    Pop-Location
+}

